sklearn/linear_model/_glm/glm.py

```diff
@@ -0,0 +1,885 @@
+"""
+Generalized Linear Models with Exponential Dispersion Family
+"""
+
+# Author: Christian Lorentzen <lorentzen.ch@googlemail.com>
+# some parts and tricks stolen from other sklearn files.
+# License: BSD 3 clause
+
+import numbers
+
+import numpy as np
+import scipy.optimize
+
+from ...base import BaseEstimator, RegressorMixin
+from ...utils import check_array, check_X_y
+from ...utils.optimize import _check_optimize_result
+from ...utils.validation import check_is_fitted, _check_sample_weight
+from .distribution import (
+        ExponentialDispersionModel,
+        TweedieDistribution,
+        EDM_DISTRIBUTIONS
+)
+from .link import (
+        Link,
+        IdentityLink,
+        LogLink,
+)
+
+
+class GeneralizedLinearRegressor(BaseEstimator, RegressorMixin):
+    """Regression via a Generalized Linear Model (GLM) with penalties.
+
+    GLMs based on a reproductive Exponential Dispersion Model (EDM) aim at
+    fitting and predicting the mean of the target y as mu=h(X*w). Therefore,
+    the fit minimizes the following objective function with L2
+    priors as regularizer::
+
+            1/(2*sum(s)) * deviance(y, h(X*w); s)
+            + 1/2 * alpha * |w|_2
+
+    with inverse link function h and s=sample_weight.
+    The parameter ``alpha`` corresponds to the lambda parameter in glmnet.
+
+    Read more in the :ref:`User Guide <Generalized_linear_regression>`.
+
+    Parameters
+    ----------
+    alpha : float, optional (default=1)
+        Constant that multiplies the penalty terms and thus determines the
+        regularization strength.
+        See the notes for the exact mathematical meaning of this
+        parameter.``alpha = 0`` is equivalent to unpenalized GLMs. In this
+        case, the design matrix X must have full column rank
+        (no collinearities).
+
+    fit_intercept : boolean, optional (default=True)
+        Specifies if a constant (a.k.a. bias or intercept) should be
+        added to the linear predictor (X*coef+intercept).
+
+    family : {'normal', 'poisson', 'gamma', 'inverse.gaussian'} \
+            or an instance of class ExponentialDispersionModel, \
+            optional(default='normal')
+        The distributional assumption of the GLM, i.e. which distribution from
+        the EDM, specifies the loss function to be minimized.
+
+    link : {'auto', 'identity', 'log'} or an instance of class Link, \
```

Comment:
> The Link class is not exposed

The Link class is internationally not part of the public API (for now), but nevertheless you can pass a Link instance here so it might as well be documented.

We have other examples of such private but documented behavior, existing for the same reasons:
 - `DecisionTreeClassifier._tree` is a `sklearn.tree._tree.Tree` object
 - `Birch._root` is a `_CFNode`

meaning that we don't guarantee backward compatibility on those, but you can use them if you know what you are doing.

The same applies to the distribution classes. Passing custom distribution and link functions can be useful in some applications, and it doesn't really cost much to support them, since that's the way the implementation is organized already.

Reply:
I'm OK with documenting the Link class as long as the `GeneralizedLinearRegressor` isn't exposed, but if we ultimately expose is, I don't think we should document the use of private API, especially for init arguments.


Regarding attributes: I would argue that we should not expose `tree_` and `root_` as public attributes if these are subject to change without deprecation. We claim that the public attributes should be backward-compatible.

Reply:
Let's continue this discussion in https://github.com/scikit-learn/scikit-learn/issues/14720